modules/library/data/src/main/java/org/geotools/data/store/ContentFeatureCollection.java

```diff
@@ -254,6 +318,24 @@ public int size() {
                    count++;
                }
                return count;
+           } else {
+               // we have to iterate, save to cache for later successive readings of data.
+               List<SimpleFeature> tempFeatureCache = new ArrayList<SimpleFeature>();
```

Comment:
Hi Jody, really the issue is caused by Geoserver code (WFS GetFeature requests), I would like to propose move the changes to geoserver.

https://github.com/geoserver/geoserver/pull/1355
Do you prefrer this fix?

Alvaro